### PR TITLE
phungminhthi/feature/t01-show-the-total-of-jobs

### PR DESCRIPTION
**Pull request by minhthiphung989**
- Call Web APIs with the useEffect Hook in React
- Use  paramaters "length" to show total jobs
- Link deploy: https://flawless-rest.surge.sh/

### CODE REVIEW

File: src/components/Dashboard/Dashboard.test.tsx [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+describe('<Dashboard />', () => {
+    it('when rendering success', () => {
+      const {container} = render(<Dashboard />);
+      expect(container.querySelectorAll('.dashBoard').length).toBe(1);
+    });
+});
+//Search jobs
+describe('Input value', () => {
+  it('updates on change', () => {
+    const setSearch = jest.fn((value) => {})
```
Comment (hienntfaba) [inline src/components/Dashboard/Dashboard.test.tsx:23]: Remove unused variables

File: src/components/Dashboard/Dashboard.test.tsx [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    const { getByLabelText } = render(<Dashboard />)
+
+    const searchInput = getByLabelText('Search') as HTMLInputElement;
+
+    fireEvent.change(searchInput, { target: { value: 'test' } })
+
+    expect(searchInput.value).toBe('test')
+  })
+  it('value is null', () => {
+    const setSearch = jest.fn((value) => {})
```
Comment (hienntfaba) [inline src/components/Dashboard/Dashboard.test.tsx:34]: Remove unused variables

File: src/components/Dashboard/Dashboard.tsx [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+const options = ['Ho Chi Minh', 'Ha Noi', 'Da Nang', 'Can Tho'];
+const Dashboard : FC = () => {
+    const [jobs, setJobs] = useState([].slice(0, 100));
+    const [pageNumber, setPageNumber] = useState(0);
+    const jobsPerPage = 10; 
+    const pagesVisited = pageNumber * jobsPerPage;
+    const [searchTerm,setSearchTerm ] = useState("");
+        
+    useEffect(() => {
+        axios.get('https://6176370c03178d00173daae3.mockapi.io/api/api')
```
Comment (hienntfaba) [inline src/components/Dashboard/Dashboard.tsx:19]: redux, redux-thunk

File: src/components/Dashboard/Dashboard.tsx [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    const changePage = ( {selected}:any ) => {
+            setPageNumber(selected);
+        };
+    
+    return (
+        <div className="dashBoard">
+            <h1>Có tất cả {jobs.length} IT Jobs For Chất Developers</h1>
+            <div className='search'>
+                <div className="search__form">
+                    <Box component="form" sx={{'& > :not(style)': { m: 1, width: '55ch' },
```
Comment (hienntfaba) [inline src/components/Dashboard/Dashboard.tsx:38]: please check alignment

File: package.json [hunk 5]
```diff
@@ -3,16 +3,32 @@
...
+    "@types/redux": "^3.6.0",
+    "@types/redux-thunk": "^2.1.0",
+    "axios": "^0.24.0",
+    "bootstrap": "5.1.3",
+    "connect-api-mocker": "^1.10.0",
     "node-sass": "^4.14.1",
-    "react": "^16.13.1",
-    "react-dom": "^16.13.1",
+    "react": "^17.0.2",
+    "react-bootstrap": "^2.0.0",
```
Comment (hienntfaba) [inline package.json:22]: Don't mix mui and react-bootstrap.

Please remove one.

File: src/components/Dashboard/Dashboard.tsx (hunk 4 above)
Comment (hienntfaba) [inline src/components/Dashboard/Dashboard.tsx:38]: sx={{'& > :not(style)': { m: 1, width: '55ch' }}

what is this used for?

File: src/containers/Home/index.tsx [hunk 6]
```diff
@@ -1,17 +1,14 @@
...
   return (
     <div className="App">
-      <header className="App-header">
-        <img src={logo} className="App-logo" alt="logo" />
-        <p>
-          Welcome to Faba onboarding project - Job board
-        </p>
-      </header>
+      <Header/>
+      {/* <Dashboard/> */}
```
Comment (hienntfaba) [inline src/containers/Home/index.tsx:11]: remove commented code

File: src/containers/Router/index.tsx [hunk 7]
```diff
@@ -1,13 +1,21 @@
...
+import AllJobs from "../../components/AllJobs/AllJobs";
 
 const Router: React.FC = () => (
-  <Switch>
-    <Route path={routes.ROOT} component={Home} />
-  </Switch>
+  <BrowserRouter>
+    <Switch>
+      <Route path={routes.ROOT} component={Home} />
+      <Route path="/dashboard" component={Dashboard} />
```
Comment (hienntfaba) [inline src/containers/Router/index.tsx:14]: please put all routes in `routes`

File: src/types/jobsType.ts [hunk 8]
```diff
@@ -0,0 +1,9 @@
+
+export interface jobs {
```
Comment (hienntfaba) [inline src/types/jobsType.ts:2]: Should be `Job`

File: package.json [hunk 9]
```diff
@@ -3,16 +3,32 @@
   "version": "0.1.0",
   "private": true,
   "dependencies": {
+    "@emotion/react": "^11.5.0",
+    "@emotion/styled": "^11.3.0",
```
Comment (hienntfaba) [inline package.json:7]: Do we use this?

**Review by hienntfaba — CHANGES_REQUESTED** [review]

File: src/components/AllJobs/AllJobs.tsx [hunk 10]
```diff
@@ -0,0 +1,39 @@
+import React, { FC, useEffect, useState } from 'react'
+import './_AllJobs.scss'
+import axios from 'axios';
+import { jobs } from '../../types/jobsType'
```
[the remark this answers is not in the document]
Comment (minhthiphung989) [reply src/components/AllJobs/AllJobs.tsx]: fixed

File: src/components/Dashboard/Dashboard.test.tsx (hunk 1 above)
Comment (minhthiphung989) [reply src/components/Dashboard/Dashboard.test.tsx:22]: fixed

File: src/containers/Router/index.tsx (hunk 7 above)
Comment (minhthiphung989) [reply src/containers/Router/index.tsx:10]: fixed

File: src/components/Dashboard/Dashboard.test.tsx (hunk 1 above)
Comment (minhthiphung989) [reply src/components/Dashboard/Dashboard.test.tsx]: fixed

File: src/components/AllJobs/AllJobs.tsx [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+    return (
+        <div>
+            <h2>All Jobs</h2>
+            {jobs.map((jobs:any) => (       
+                <div key={jobs.id} className="jobsCard">
+                    <div className="jobsCard__left">
+                        <img src={jobs.jobImg} alt="" />
+                    </div>
+                    <div className = "jobsCard__center">
+                        <a href="" id="name">{jobs.jobName}</a>
```
[the remark this answers is not in the document]
Comment (minhthiphung989) [reply src/components/AllJobs/AllJobs.tsx]: fixed

File: src/components/Dashboard/Dashboard.tsx (hunk 4 above)
Comment (minhthiphung989) [reply src/components/Dashboard/Dashboard.tsx]: fixed

File: src/containers/Home/index.tsx (hunk 6 above)
Comment (minhthiphung989) [reply src/containers/Home/index.tsx]: fixed

File: package.json (hunk 9 above)
Comment (minhthiphung989) [reply package.json]: fixed

File: src/components/Dashboard/Dashboard.tsx (hunk 4 above)
Comment (minhthiphung989) [reply src/components/Dashboard/Dashboard.tsx]: fixed

File: package.json (hunk 5 above)
Comment (minhthiphung989) [reply package.json:19]: fixed

File: src/components/Dashboard/Dashboard.tsx [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+
+                        <Dropdown.Menu>
+                            <Dropdown.Item href="#/action-1">Ha Noi</Dropdown.Item>
+                            <Dropdown.Item href="#/action-2">Ho Chi Minh</Dropdown.Item>
+                            <Dropdown.Item href="#/action-3">Da Nang</Dropdown.Item>
+                        </Dropdown.Menu>
+                    </Dropdown>
+                </div>
+            </div>
+            {jobs.filter((val: any) => {
```
Comment (hienntfaba) [inline src/components/Dashboard/Dashboard.tsx:52]: don't use `any`

Comment (hienntfaba) [inline src/components/Dashboard/Dashboard.tsx:52]: try moving filter and slice into reducer

File: src/components/Header.tsx [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+                                Employers
+                            </Nav.Link>
+                        </Nav>
+                    </Navbar.Collapse>
+                </Container>
+            </Navbar>
+            <div className="container mt-3">
+                <BrowserRouter>
+                    <Switch>
+                        <Route exact path={["/"]} component={Dashboard} />
```
Comment (hienntfaba) [inline src/components/Header.tsx:38]: use `routes` constant

File: src/containers/Home/Home.test.tsx [hunk 14]
```diff
@@ -1,9 +1,8 @@
-import {render} from "@testing-library/react";
-import Home from "../Home/Home";
-import React from "react";
-
+import { render } from "@testing-library/react";
+import Header from "../../components/Header";
+import { Provider } from 'react-redux';
+import store from '../../store'
 test('renders learn react link', () => {
```
Comment (hienntfaba) [inline src/containers/Home/Home.test.tsx:5]: update description

File: src/containers/Router/services/index.ts [hunk 15]
```diff
@@ -1,3 +1,5 @@
 export const routes = {
   ROOT: "/",
+  dashBoard: "/dasboard",
```
Comment (hienntfaba) [inline src/containers/Router/services/index.ts:3]: constant should be UPPERCASE

`DASHBOARD`

File: src/reducer/reducer.tsx [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+            return {
+                ...state,
+                loading: false,
+                Job: action.payload
+            }
+        case FETCH_JOB_FAIL:
+            return {
+                ...state,
+                loading: false,
+                Job: action.error
```
Comment (hienntfaba) [inline src/reducer/reducer.tsx:30]: why assign error to Job?

File: src/components/Dashboard/__snapshots__/Dashboard.test.tsx.snap [hunk 17]
```diff
@@ -0,0 +1,167 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
```
Comment (hienntfaba) [inline src/components/Dashboard/__snapshots__/Dashboard.test.tsx.snap:1]: Update snapshot

File: src/components/Header.tsx [hunk 18]
```diff
@@ -0,0 +1,48 @@
+import React from 'react'
+import { Switch, Route, Link, BrowserRouter } from "react-router-dom";
+import { Navbar, Container, Nav, NavDropdown, Collapse } from 'react-bootstrap';
```
Comment (nhatphfaba) [inline src/components/Header.tsx:3]: Remove unused import on this file

File: src/components/Header.tsx [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+    return (
+        <div className='header'>
+            <Navbar collapseOnSelect expand="lg" bg="dark" variant="dark">
+                <Container>
+                    <Navbar.Brand href="/">ITviec</Navbar.Brand>
+                    <Navbar.Toggle aria-controls="responsive-navbar-nav" />
+                    <Navbar.Collapse id="responsive-navbar-nav">
+                        <Nav className="me-auto">
+                            <Nav.Link href="/alljobs">All Jobs</Nav.Link>
+                            <NavDropdown title="IT Companies" id="collasible-nav-dropdown">
```
Comment (nhatphfaba) [inline src/components/Header.tsx:17]: Remove all unused Nav component  on this file.

**Review by nhatphfaba — CHANGES_REQUESTED** [review]

File: src/reducer/reducer.tsx [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+    jobs: Job[],
+    allJobs: Job[],
+    loading: boolean,
+    error: null,
+    value: string,
+    perpage: number;
+    pagecount: number;
+    offset: number;
+}
+export function jobsReducer(state = initialStare, action: any) {
```
Comment (nhatphfaba) [inline src/reducer/reducer.tsx:23]: please declare the type of this state

**Review by nhatphfaba — CHANGES_REQUESTED** [review]

File: src/data/api.ts [hunk 21]
```diff
@@ -0,0 +1,20 @@
+import { fetchJobsLoading, fetchJobsSuccess, fetchJobsError } from '../action/actions'
+import { Dispatch } from 'redux';
+export function fetchJobs() {
+    return (dispatch: Dispatch) => {
```
Comment (nhatphfaba) [inline src/data/api.ts:4]: why does this fetch 2 times on load? there should be only one fetch.

**Review by nhatphfaba — CHANGES_REQUESTED** [review]

**Pull request merged**